spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/ConverterAwareMappingDatastoreEntityConverter.java

```diff
@@ -0,0 +1,194 @@
+			customConverters.forEach(this.conversionService::addConverter);
+		}
+		DEFAULT_CONVERTERS.forEach(this.conversionService::addConverter);
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public <R> R read(Class<R> aClass, Entity entity) {
+		DatastorePersistentEntity<R> persistentEntity =
+				(DatastorePersistentEntity<R>) this.mappingContext.getPersistentEntity(aClass);
+
+		EntityPropertyValueProvider propertyValueProvider =
+				new EntityPropertyValueProvider(entity, this.conversionService, this.customConversions);
+
+		ParameterValueProvider<DatastorePersistentProperty> parameterValueProvider =
+				new PersistentEntityParameterValueProvider<>(persistentEntity, propertyValueProvider, null);
+
+		EntityInstantiator instantiator = this.instantiators.getInstantiatorFor(persistentEntity);
+		R instance = instantiator.createInstance(persistentEntity, parameterValueProvider);
+		PersistentPropertyAccessor accessor = persistentEntity.getPropertyAccessor(instance);
+
+		persistentEntity.doWithProperties(
+				(PropertyHandler<DatastorePersistentProperty>) datastorePersistentProperty -> {
+					Object value = propertyValueProvider.getPropertyValue(datastorePersistentProperty);
+					accessor.setProperty(datastorePersistentProperty, value);
+				});
+
+		return instance;
+	}
+
+	@Override
+	public void write(Object source, Entity.Builder sink) {
+		DatastorePersistentEntity<?> persistentEntity = this.mappingContext.getPersistentEntity(source.getClass());
+		PersistentPropertyAccessor accessor = persistentEntity.getPropertyAccessor(source);
+		persistentEntity.doWithProperties(
+				(DatastorePersistentProperty datastorePersistentProperty) ->
+						writeProperty(sink, accessor, datastorePersistentProperty));
+	}
+
+	@SuppressWarnings("unchecked")
+	private void writeProperty(Entity.Builder sink, PersistentPropertyAccessor accessor,
+			DatastorePersistentProperty persistentProperty) {
+		Object propertyVal = convertToSimpleType(accessor.getProperty(persistentProperty), persistentProperty);
+		if (propertyVal == null) {
+			sink.set(persistentProperty.getFieldName(), new NullValue());
+			return;
+		}
+		Optional<Class<?>> basicTargetType = this.toNativeConversions.getCustomWriteTarget(propertyVal.getClass());
+		if (basicTargetType.isPresent()) {
+			propertyVal = this.toNativeConversionService.convert(propertyVal, basicTargetType.get());
+		}
+		Function valueFactory = VAL_FACTORIES.get(propertyVal.getClass());
+
+		if (valueFactory == null) {
+			throw new DatastoreDataException("Simple type value factory is not found for property" +
+					" with name " + persistentProperty.getFieldName()
+					+ ". The property's simple type is " + propertyVal.getClass());
+		}
+
+		Value val = (Value) valueFactory.apply(propertyVal);
+		sink.set(persistentProperty.getFieldName(), val);
+	}
+
+	private Object convertToSimpleType(@Nullable Object obj, DatastorePersistentProperty prop) {
+		if (obj == null || DatastoreSimpleTypes.isSimple(obj.getClass())) {
```

Comment:
You only need to check for `null`, and rest should be taken care of by `getCustomWriteTarget`.